pkg/reconciler/extension/extension.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2022 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package extension
+
+import (
+	"context"
+
+	autoscalingv1alpha1 "knative.dev/serving/pkg/apis/autoscaling/v1alpha1"
+	v1 "knative.dev/serving/pkg/apis/serving/v1"
+)
+
+// ServingExtension enables pluggable extensive features added into the controller
+type ServingExtension interface {
+	// TransformRevision transforms the existing revision.
+	TransformRevision(*v1.Revision) *v1.Revision
+	// PostRevisionReconcile implements a custom reconcile loop the after the revision is created.
+	PostRevisionReconcile(context.Context, *autoscalingv1alpha1.PodAutoscaler, *autoscalingv1alpha1.PodAutoscaler) error
```

Comment:
What's the purpose of these `Post*Reconcile` functions?


If you're compiling your own revision controllers (which surprises me) then you don't need these hooks cause you can just wrap the existing reconciler with your own

https://github.com/knative/serving/blob/ea603c9685fb09d46793477b798ae72c7d88a7d8/pkg/reconciler/service/controller.go#L52-L57

Reply:
I am thinking of adding these hooks inside the serving code, so that in the extension, I only need to implement the interface `ServingExtension` to accomplish the progressive rollout.

PostConfigurationReconcile: I will leverage this function to create the extension's CR.
PostRevisionReconcile: I just use this function to update the podautoscaler, so it only picks up the minScale and maxScale based on the current rolling stage.